README.md
# flutter_wordpress_content

A Flutter helper widget to parse WordPress content in raw form.

## Getting Started

In order for this plugin to work, you must provide the raw content which is saved
in the database (not the rendered content returned by the API). For example see below 
comparison:

###### Raw Content
```$xslt
<!-- wp:paragraph -->Some text to be shown<!-- /wp:paragraph -->
<!-- wp:image --><figure><img src="https://example.com/image.png" /></figure><!-- /wp:image -->
```
###### Rendered Content
```$xslt
<p>Some text to be shown</p>
<img src="https://example.com/image.png" />
```
---
### Using WPContent Class
For detail on providing YouTube, SoundCloud or Issue widgets, see __Supported Types__ section below.
```dart
Widget buildMyContent(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: SingleChildScrollView(
          child: WPContent(
            "raw_content",
            headingTextColor: Colors.black,
            paragraphTextColor: Colors.black,
            imageCaptionTextColor: Colors.black,
            textDirection: TextDirection.ltr,
            fontFamily: 'my_font_family',
            fontSize: 19.0,
            paragraphArabicIdentifier: 'tk-adobe-arabic',
            arabicFontFamily: 'my_arabic_font_family',
            youtubeEmbedWidget: YouTubeEmbedWidget(),
            soundcloudEmbedWidget:
                SoundCloudEmbedWidget("Audio Title", "Audio Subtitle"),
            issuuEmbedWidget: IssueEmbedWidget(),
            jwPlayerWidget: JWPlayerEmbedWidget(),
          ),
        ),
      ),
    );
  }
```
### Supported Types
At this moment, this plugin supports the following types of content.

#### Heading `<!-- wp:heading -->`
Standard WP heading.

#### Paragraph `<!-- wp:paragraph -->`
Standard WP paragraph.

#### YouTube Embed `<!-- wp:core-embed/youtube -->`
For embedding YouTube, create your own widget for displaying YouTube video and inherit
from the provided `YouTubeWidget`. This widget will give you access to the `videoId`. Use
this ID in your widget to render YouTube player.
```dart
class YouTubeEmbedWidget extends YouTubeWidget {
  @override
  Widget buildWithVideoId(BuildContext context, String videoId) {
    return Container(
      padding: EdgeInsets.fromLTRB(0.0, 10.0, 0.0, 20.0),
      child: AspectRatio(
        aspectRatio: 16 / 9,
        child: Container(
          child: Text(videoId),
        ),
      ),
    );
  }
}
```

#### Issue PDF Embed `<!-- wp:core-embed/issuu -->`
For embedding Issuu pdfs, create your own widget for displaying PDF and inherit
from the provided `IssuuWidget`. This widget will give you access to an instance of `SimpleArticle`.
```dart
class IssueEmbedWidget extends IssuuWidget {
  @override
  Widget buildWithPDF(BuildContext context, SimpleArticle pdf) {
    return Container(
      padding: EdgeInsets.all(10.0),
      child: Center(
        child: RaisedButton(
            padding: EdgeInsets.all(10.0),
            color: Colors.green,
            child: Container(
              padding: EdgeInsets.only(top: 5.0),
              child: Text(
                "View PDF",
                style: Theme.of(context)
                    .textTheme
                    .body1
                    .copyWith(color: Colors.white),
              ),
            ),
            onPressed: () {
              Navigator.push(context, MaterialPageRoute<void>(
                builder: (context) {
                  return Container(
                    child: Text(pdf.paragraphRawContent),
                  );
                },
              ));
            }),
      ),
    );
  }
}

```
The instance of `SimpleArticle` will give you access to raw content where you can extract
the PDF url. See below for an example. Use this URL to then display PDF in your widget.
```dart
String rawContent = mySimpleArticle.paragraphRawContent;

String extractedURL =
  RegExp(r'(?:"url":")(.*?)(?:")').firstMatch(rawContent).group(1);
```

#### SoundCloud Embed `<!-- wp:html -->`
When embedding SoundCloud, please make sure you use the embed code provided by
SoundCloud and not the standard way of embedding in WordPress. This is necessary
because the embed code contains track ID which is used to fetch more information
about the track through SC API.

Inheriting from the provided `SoundCloudWidget` will give you access to `trackId` field.
Use this ID to fetch the stream URL and metadata and play the audio. For an example of
how to do that, please see the [flutter_playout](https://pub.dev/packages/flutter_playout) plugin.
```dart
class SoundCloudEmbedWidget extends SoundCloudWidget {
  final String title;
  final String subtitle;

  SoundCloudEmbedWidget(this.title, this.subtitle);

  @override
  Widget buildWithTrackId(BuildContext context, String trackId, String embedCode) {
    return Container(
      padding: EdgeInsets.fromLTRB(0.0, 10.0, 0.0, 30.0),
      child: Container(
        child: Column(
          children: <Widget>[Text(title), Text(subtitle), Text(trackId)],
        ),
      ),
    );
  }
}
```

#### HearthisAt (hearthis.at) Embed `<!-- wp:html -->`
When embedding audios from hearthis.at, please make sure you use the embed code 
from the share button on the track.

Inheriting from the provided `HearthisAtWidget` will give you access to `trackId` field.
Use this ID to fetch the stream URL and metadata and play the audio. For an example of
how to do that, please see the [flutter_playout](https://pub.dev/packages/flutter_playout) plugin.
```dart
class HearthisAtEmbedWidget extends HearthisAtWidget {
  final String title;
  final String subtitle;

  HearthisAtEmbedWidget(this.title, this.subtitle);

  @override
  Widget buildWithTrackId(BuildContext context, String trackId) {
    return Container(
      padding: EdgeInsets.fromLTRB(0.0, 10.0, 0.0, 30.0),
      child: Container(
        child: Column(
          children: <Widget>[Text(title), Text(subtitle), Text(trackId)],
        ),
      ),
    );
  }
}
```

#### JWPlayer Embed
For embedding JWPlayer hosted media, create your own widget for displaying video and inherit
from the provided `JWPlayerWidget`. This widget will give you access to the `mediaId`. Use
this ID in your widget to render the player.
```dart
class JWPlayerEmbedWidget extends JWPlayerWidget {
  @override
  Widget buildWithMediaId(BuildContext context, String mediaId) {
    return Container(
      padding: EdgeInsets.fromLTRB(0.0, 10.0, 0.0, 20.0),
      child: AspectRatio(
        aspectRatio: 16 / 9,
        child: Container(
          child: Text(mediaId),
        ),
      ),
    );
  }
}
```

CHANGELOG.md
## [1.0.26] - January 20, 2022

* Fixes an issue causing nested text spans to not be tap-able

## [1.0.25] - October 22, 2021

* Updated to Android v2 Embedding API
* Using null-safety

## [1.0.24] - April 1, 2020

* Added categoryId and parentCategoryId to model/SimpleArticle.dart

## [1.0.23] - 7 March, 2020

* added ability to change default text alignment for paragraph text

## [1.0.22] - 2 March, 2020

* Fixed an issue where multiple <strong> tags were not being parsed

## [1.0.21] - 22 February, 2020

* Updated Hearthis.at base url to app.hearthis.at for embeds

## [1.0.20] - 8 February, 2020

* Added ability to change text alignment and colour for WP Quotes independent of paragraph style

## [1.0.19] - 4 February, 2020

* Added support for WP List (<!-- wp:list -->)

## [1.0.18] - 4 February, 2020

* Added support for WP Quotes (<!-- wp:quote -->)

## [1.0.17] - 2 January, 2020

* Fixed an issue causing heading to be right aligned in ltr alignment

## [1.0.16] - 10 December, 2019

* Upgraded cached_network_image to 2.0.0-rc

## [1.0.15] - 10 December, 2019

* Added HearthisAtWidget for embedding audios hosted by your [hearthis.at](https://hearthis.at) account

## [1.0.14] - 27 November, 2019

* Changed SoundCloudEmbedWidget to also include original embed code for SoundCloud widget

## [1.0.13] - 6 October, 2019

* Override == operator in `SimpleArticle` so that it can work in `contains(object)` in Maps and Lists

## [1.0.12] - 6 October, 2019

* Added better exception handling for `SimpleArticle`'s to and from json methods

## [1.0.11] - 5 October, 2019

* Added `FeatureImage` class replacing `imageUrl` field in `SimpleArticle`
* Added the ability to change text colour for Heading, Paragraph and Image Caption

## [1.0.10] - 2 October, 2019

* added parser for <a> tags with url_launcher to open links in browser

## [1.0.9] - 2 October, 2019

* Implemented Superscript <sup>, Subscript <sub> and <strong> tag parser

## [1.0.8] - 2 October, 2019

* Fixed an issue with image captions where super/sub scripts weren't being parsed

## [1.0.7] - 2 October, 2019

* Fixed an issue with TextSpan rendering

## [1.0.6] - 2 October, 2019

* Updated wp_parser to use TextSpan with RichText instead of Unicode for superscript and subscript

## [1.0.5] - 2 October, 2019

* Added JWPlayerWidget for embedding media hosted by your JW Player Account

## [1.0.4] - 2 October, 2019

* Fixed an issue with paragraph spacing

## [1.0.3] - 2 October, 2019

* Updated package description

## [1.0.2] - 2 October, 2019

* Changed embed helper widgets to simple classes

## [1.0.1] - 2 October, 2019

* Changed embed helper widgets to StatelessWidgets
* Added example implementation in documentation

## [1.0.0] - 1 October, 2019

* Added content parser for Heading, Paragraph, YouTube, SoundCloud & Issuu embeds.


test/flutter_wordpress_content_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_wordpress_content/external/HearthisAtWidget.dart';
import 'package:flutter_wordpress_content/external/SoundCloudWidget.dart';

void main() {
  testWidgets('Extract shortcode key value with regex',
      (WidgetTester tester) async {
    String c = "shortcode -->\n[jwplayer HxVE7bbK]\n<!-- /wp:shortcode -->";

    RegExpMatch? regExMatch = RegExp(r'\[(\w+[\s*]+)*(\w+)\]').firstMatch(c);

    try {
      if (regExMatch != null) {
        String? shortcodeType = regExMatch.group(1);

        if (shortcodeType != null) {
          switch (shortcodeType.trim()) {
            /* jwplayer */
            case "jwplayer":
              String? mediaId = regExMatch.group(2);

              if (mediaId != "HxVE7bbK") {
                throw new AssertionError("media id mismatch");
              }
              break;
            default:
              break;
          }
        }
      }
    } catch (exception) {/* ignore */}
  });

  testWidgets('Extract trackId from hearthis.at embed',
      (WidgetTester tester) async {
    String c =
        "html -->\n<figure><iframe width=\"100%\" height=\"150\" src=\"https://hearthis.at/embed/4147030/transparent_black/?hcolor=&amp;color=&amp;style=2&amp;block_size=2&amp;block_space=1&amp;background=1&amp;waveform=0&amp;cover=0&amp;autoplay=0&amp;css=\"></iframe></figure>\n<!-- /wp:html -->";

    String trackId = HearthisAtWidget.getIdFromUrl(c);

    assert(trackId == "4147030");
  });

  testWidgets('Extract trackId from SoundCloud embed',
      (WidgetTester tester) async {
    String c =
        "html -->\n<figure><iframe width=\"100%\" height=\"20\" src=\"https://w.soundcloud.com/player/?url=https%3A//api.soundcloud.com/tracks/725625592&amp;color=%23ff5500&amp;inverse=false&amp;auto_play=true&amp;show_user=true\"></iframe></figure>\n<!-- /wp:html -->";

    String trackId = SoundCloudWidget.getIdFromUrl(c);

    assert(trackId == "725625592");
  });
}


example/lib/main.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter_wordpress_content/external/FlipHTML5Widget.dart';
import 'package:flutter_wordpress_content/external/HearthisAtWidget.dart';
import 'package:flutter_wordpress_content/external/IssuuWidget.dart';
import 'package:flutter_wordpress_content/external/JWPlayerWidget.dart';
import 'package:flutter_wordpress_content/external/SoundCloudWidget.dart';
import 'package:flutter_wordpress_content/external/YouTubeWidget.dart';
import 'package:flutter_wordpress_content/model/SimpleArticle.dart';
import 'package:flutter_wordpress_content/wp_content.dart';
import 'package:url_launcher/url_launcher.dart';

void main() => runApp(WordPressContentExample());

class WordPressContentExample extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(),
        body: SingleChildScrollView(
          child: WPContent(
            '<!-- wp:paragraph --><a href="https://www.mta.tv">MTA</a> hello <!-- /wp:paragraph -->',
            headingTextColor: Colors.black,
            paragraphTextColor: Colors.black,
            imageCaptionTextColor: Colors.black,
            textDirection: TextDirection.ltr,
            fontFamily: 'my_font_family',
            fontSize: 19.0,
            paragraphArabicIdentifier: 'tk-adobe-arabic',
            arabicFontFamily: 'my_arabic_font_family',
            youtubeEmbedWidget: YouTubeEmbedWidget(),
            soundcloudEmbedWidget:
                SoundCloudEmbedWidget("Audio Title", "Audio Subtitle"),
            hearthisAtWidget:
                HearthisAtEmbedWidget("Audio Title", "Audio Subtitle"),
            issuuEmbedWidget: IssueEmbedWidget(),
            jwPlayerWidget: JWPlayerEmbedWidget(),
          ),
        ),
      ),
    );
  }
}

class Tester extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.fromLTRB(0.0, 16.0, 0.0, 5.0),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.start,
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[
          Container(
            padding: EdgeInsets.fromLTRB(10.0, 0.0, 10.0, 13.0),
            child: RichText(
              text: TextSpan(
                children: [
                  TextSpan(
                      text: "MTA",
                      style: TextStyle(color: Colors.blue[800]),
                      recognizer: new TapGestureRecognizer()
                        ..onTap = () async {
                          if (await canLaunch("https://www.mta.tv")) {
                            await launch("https://www.mta.tv");
                          }
                        })
                ],
                style: TextStyle(color: Colors.blue[800]),
              ),
            ),
          )
        ],
      ),
    );
  }
}

class YouTubeEmbedWidget extends YouTubeWidget {
  @override
  Widget buildWithVideoId(BuildContext context, String videoId) {
    return Container(
      padding: EdgeInsets.fromLTRB(0.0, 10.0, 0.0, 20.0),
      child: AspectRatio(
        aspectRatio: 16 / 9,
        child: Container(
          child: Text(videoId),
        ),
      ),
    );
  }
}

class SoundCloudEmbedWidget extends SoundCloudWidget {
  final String title;
  final String subtitle;

  SoundCloudEmbedWidget(this.title, this.subtitle);

  @override
  Widget buildWithTrackId(
      BuildContext context, String trackId, String embedCode) {
    return Container(
      padding: EdgeInsets.fromLTRB(0.0, 10.0, 0.0, 30.0),
      child: Container(
        child: Column(
          children: <Widget>[Text(title), Text(subtitle), Text(trackId)],
        ),
      ),
    );
  }
}

class HearthisAtEmbedWidget extends HearthisAtWidget {
  final String title;
  final String subtitle;

  HearthisAtEmbedWidget(this.title, this.subtitle);

  @override
  Widget buildWithTrackId(BuildContext context, String trackId) {
    return Container(
      padding: EdgeInsets.fromLTRB(0.0, 10.0, 0.0, 30.0),
      child: Container(
        child: Column(
          children: <Widget>[Text(title), Text(subtitle), Text(trackId)],
        ),
      ),
    );
  }
}

class IssueEmbedWidget extends IssuuWidget {
  @override
  Widget buildWithPDF(BuildContext context, SimpleArticle pdf) {
    return Container(
      padding: EdgeInsets.all(10.0),
      child: Center(
        child: RaisedButton(
            padding: EdgeInsets.all(10.0),
            color: Colors.green,
            child: Container(
              padding: EdgeInsets.only(top: 5.0),
              child: Text(
                "View PDF",
                style: Theme.of(context)
                    .textTheme
                    .bodyText2
                    .copyWith(color: Colors.white),
              ),
            ),
            onPressed: () {
              Navigator.push(context, MaterialPageRoute<void>(
                builder: (context) {
                  return Container(
                    child: Text(pdf.paragraphRawContent),
                  );
                },
              ));
            }),
      ),
    );
  }
}

class JWPlayerEmbedWidget extends JWPlayerWidget {
  @override
  Widget buildWithMediaId(BuildContext context, String mediaId) {
    return Container(
      padding: EdgeInsets.fromLTRB(0.0, 10.0, 0.0, 20.0),
      child: AspectRatio(
        aspectRatio: 16 / 9,
        child: Container(
          child: Text(mediaId),
        ),
      ),
    );
  }
}

class FlipHTML5EmbedWidget extends FlipHTML5Widget {
  @override
  Widget buildWithPDF(BuildContext context, SimpleArticle pdf) {
    return Container(
      padding: EdgeInsets.all(10.0),
      child: Center(
        child: RaisedButton(
            padding: EdgeInsets.all(10.0),
            color: Colors.green,
            child: Container(
              padding: EdgeInsets.only(top: 5.0),
              child: Text(
                "View PDF",
                style: Theme.of(context)
                    .textTheme
                    .bodyText2
                    .copyWith(color: Colors.white),
              ),
            ),
            onPressed: () {
              Navigator.push(context, MaterialPageRoute<void>(
                builder: (context) {
                  return Container(
                    child: Text(pdf.paragraphRawContent),
                  );
                },
              ));
            }),
      ),
    );
  }
}


lib/wp_content.dart
library flutter_wordpress;

import 'dart:ui';

import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:flutter_wordpress_content/external/HearthisAtWidget.dart';
import 'package:flutter_wordpress_content/external/JWPlayerWidget.dart';

import 'external/IssuuWidget.dart';
import 'external/SoundCloudWidget.dart';
import 'external/YouTubeWidget.dart';
import 'model/Paragraph.dart';
import 'model/SimpleArticle.dart';
import 'wp_parser.dart';

class WPContent extends StatelessWidget {
  // raw WordPress content (not rendered). This content should be
  // in <!-- wp: --> tags such as for paragraph <!-- wp:paragraph -->
  // or for image <!-- wp:image -->
  final String rawWPContent;

  // set text direction
  final TextDirection textDirection;

  // set colour for the heading text
  final Color headingTextColor;

  // set colour for the paragraph text
  final Color paragraphTextColor;

  // set colour for the image caption text
  final Color imageCaptionTextColor;

  // default font family for text
  final String fontFamily;

  // default font size for text
  final double fontSize;

  // if a paragraph contains arabic then provide an identifier
  // to detect arabic for example
  // <!-- wp:paragraph {"align":"center","className":"tk-adobe-arabic"}
  // here we can set the paragraphArabicIdentifier to 'tk-adobe-arabic'
  // as this will be set on paragraphs containing arabic text. Please note
  // that if this identifier appears in normal text anywhere in the paragraph,
  // it might cause the paragraph to be considered as arabic text as well.
  final String paragraphArabicIdentifier;

  // font family for displaying arabic text
  final String arabicFontFamily;

  // default text alignment for paragraph text if none specified in wp-paragraph
  // tag
  final TextAlign defaultParagraphTextAlign;

  // text alignment for WP Quote <!-- wp:quote -->
  final TextAlign quoteTextAlignment;

  // text colour for WP Quote <!-- wp:quote -->
  final Color quoteTextColour;

  // provide a widget to display YouTube embedded videos
  final YouTubeWidget youtubeEmbedWidget;

  // provide a widget to display SoundCloud embedded audios
  final SoundCloudWidget soundcloudEmbedWidget;

  // provide a widget to display Hearthis.at embedded audios
  final HearthisAtWidget hearthisAtWidget;

  // provide a widget to display Issuu embedded PDFs
  final IssuuWidget issuuEmbedWidget;

  // provide a widget to display JWPlayer video
  final JWPlayerWidget jwPlayerWidget;

  const WPContent(this.rawWPContent,
      {this.textDirection = TextDirection.ltr,
      this.headingTextColor = Colors.black,
      this.paragraphTextColor = Colors.black,
      this.imageCaptionTextColor = Colors.black,
      this.fontFamily = '',
      this.fontSize = 16.0,
      this.paragraphArabicIdentifier = "",
      this.arabicFontFamily = '',
      this.defaultParagraphTextAlign = TextAlign.justify,
      this.quoteTextAlignment = TextAlign.center,
      this.quoteTextColour = Colors.black,
      this.youtubeEmbedWidget = const YouTubeWidget(),
      this.soundcloudEmbedWidget = const SoundCloudWidget(),
      this.hearthisAtWidget = const HearthisAtWidget(),
      this.issuuEmbedWidget = const IssuuWidget(),
      this.jwPlayerWidget = const JWPlayerWidget()});

  @override
  Widget build(BuildContext context) {
    return _buildParagraphs(context, rawWPContent);
  }

  Widget _buildParagraphs(BuildContext context, String rawContent) {
    List<Paragraph> paragraphs = _processParagraphs(rawContent);

    if (paragraphs == null || paragraphs.length < 1) {
      return Container();
    }

    return Container(
      padding: EdgeInsets.fromLTRB(0.0, 16.0, 0.0, 5.0),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.start,
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisSize: MainAxisSize.min,
        children: paragraphs
            .map((paragraph) => _buildParagraph(context, paragraph))
            .toList(),
      ),
    );
  }

  Widget _buildParagraph(BuildContext context, Paragraph paragraph) {
    Alignment alignment = Alignment.centerLeft;
    if (paragraph.textAlign == TextAlign.left) {
      alignment = Alignment.centerLeft;
    } else if (paragraph.textAlign == TextAlign.center) {
      alignment = Alignment.center;
    } else if (paragraph.textAlign == TextAlign.right) {
      alignment = Alignment.centerRight;
    }

    /* paragraph - heading */
    if (paragraph.type == "heading") {
      return Container(
        padding: EdgeInsets.fromLTRB(10.0, 0.0, 10.0, 19.0),
        alignment: alignment,
        child: RichText(
          text: TextSpan(
            children: paragraph.textSpans,
            style: DefaultTextStyle.of(context).style.copyWith(
                color: headingTextColor,
                fontFamily: paragraph.fontFamily,
                fontSize: fontSize + 5.0),
          ),
          textDirection: textDirection,
          textAlign: paragraph.textAlign,
        ),
      );
    }

    /* paragraph - text */
    else if (paragraph.type == "text") {
      return Container(
        padding: EdgeInsets.fromLTRB(10.0, 0.0, 10.0, 13.0),
        alignment: alignment,
        child: RichText(
          text: TextSpan(
            children: paragraph.textSpans,
            style: DefaultTextStyle.of(context).style.copyWith(
                color: paragraphTextColor,
                fontFamily: paragraph.fontFamily,
                fontSize: fontSize),
          ),
          textDirection: textDirection,
          textAlign: paragraph.textAlign,
        ),
      );
    }

    /* paragraph - quote */
    else if (paragraph.type == "quote") {
      return Container(
        padding: EdgeInsets.fromLTRB(25.0, 0.0, 25.0, 13.0),
        alignment: alignment,
        child: RichText(
          text: TextSpan(
            children: paragraph.textSpans,
            style: DefaultTextStyle.of(context).style.copyWith(
                color: quoteTextColour,
                fontFamily: paragraph.fontFamily,
                fontSize: fontSize,
                fontWeight: FontWeight.bold),
          ),
          textDirection: textDirection,
          textAlign: paragraph.textAlign,
        ),
      );
    }

    /* paragraph - image */
    else if (paragraph.type == "image") {
      return Container(
        padding: EdgeInsets.fromLTRB(3.0, 10.0, 3.0, 20.0),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisSize: MainAxisSize.max,
          children: <Widget>[
            Thumbnail(
              paragraph.imageUri,
              Colors.transparent,
              5.0,
            ),
            Container(
              height: 7.0,
            ),
            /* image caption */
            RichText(
              text: TextSpan(
                children: paragraph.textSpans,
                style: DefaultTextStyle.of(context).style.copyWith(
                    color: imageCaptionTextColor,
                    fontFamily: fontFamily,
                    fontSize: 0.7 * fontSize),
              ),
            )
          ],
        ),
      );
    }

    /* paragraph - jwplayer embed (https://jwplayer.com) */
    else if (paragraph.type == "jwplayer" && jwPlayerWidget != null) {
      return jwPlayerWidget.buildWithMediaId(context, paragraph.jwMediaId);
    }

    /* paragraph - youtube embed (https://youtube.com) */
    else if (paragraph.type == "youtube" && youtubeEmbedWidget != null) {
      return youtubeEmbedWidget.buildWithVideoId(
          context, paragraph.youtubeVideoId);
    }

    /* paragraph - soundcloud embed (https://soundcloud.com) */
    else if (paragraph.type == "soundcloud" && soundcloudEmbedWidget != null) {
      return soundcloudEmbedWidget.buildWithTrackId(
          context, paragraph.soundcloudTrackId, paragraph.soudcloudEmbedCode);
    }

    /* paragraph - hearthis.at embed (https://hearthis.at) */
    else if (paragraph.type == "hearthis.at" && hearthisAtWidget != null) {
      return hearthisAtWidget.buildWithTrackId(
          context, paragraph.hearthisAtTrackId);
    }

    /* paragraph - issuu embed (http://issuu.com) */
    else if (paragraph.type == "issuu" && issuuEmbedWidget != null) {
      return issuuEmbedWidget.buildWithPDF(context, paragraph.pdf);
    }

    /* paragraph - unexpected (silently ignore) */
    else {
      return Container();
    }
  }

  List<Paragraph> _processParagraphs(String rawContent) {
    List<Paragraph> processedParagraphs = [];
    try {
      List<String> contentParts = rawContent.split("<!-- wp:");

      contentParts.forEach((c) {
        if (c.startsWith("image")) {
          String caption = "";

          if (c.contains("<figcaption>")) {
            caption = c.substring(
                c.indexOf("<figcaption>") + 12, c.indexOf("</figcaption>"));
          }

          RegExpMatch? fm = RegExp(r'src\s*=\s*"(.+?)"').firstMatch(c);
          processedParagraphs.add(Paragraph.image(
              fm != null ? fm.group(1) : "",
              caption,
              parseFigureCaptionHTML(caption, baseFontSize: 0.7 * fontSize)));
        } else if (c.startsWith("core-embed/issuu")) {
          processedParagraphs
              .add(Paragraph.issuu(SimpleArticle.pdfArticle(c, null)));
        } else if (c.startsWith("shortcode")) {
          RegExpMatch? regExMatch =
              RegExp(r'\[(\w+[\s*]+)*(\w+)\]').firstMatch(c);

          try {
            if (regExMatch != null) {
              switch (regExMatch.group(1)!.trim()) {
                case "jwplayer":
                  processedParagraphs
                      .add(Paragraph.jwplayer(regExMatch.group(2)));
                  break;
                default:
                  break;
              }
            }
          } catch (exception) {/* ignore */}
        } else if (c.startsWith("core-embed/youtube")) {
          processedParagraphs
              .add(Paragraph.youtubeEmbed(YouTubeWidget.getIdFromUrl(c)));
        } else if (c.startsWith("html")) {
          /* soundcloud - embed */
          if (c.contains("soundcloud.com")) {
            var soundcloudTrackId = SoundCloudWidget.getIdFromUrl(c);
            var soundcloudEmbedCode = c.substring(
                c.indexOf("-->") + 3, c.lastIndexOf("<!-- /wp:html -->"));
            if (soundcloudTrackId != null) {
              processedParagraphs.add(Paragraph.soundcloudEmbed(
                  soundcloudTrackId, soundcloudEmbedCode));
            }
          }

          /* hearthis.at - embed */
          else if (c.contains("hearthis.at")) {
            var hearthisAtTrackId = HearthisAtWidget.getIdFromUrl(c);
            if (hearthisAtTrackId != null) {
              processedParagraphs
                  .add(Paragraph.hearthisAtEmbed(hearthisAtTrackId));
            }
          }
        } else if (c.startsWith("heading")) {
          String headingContent = c
              .substring(
                  c.indexOf("-->") + 3, c.lastIndexOf("<!-- /wp:heading -->"))
              .replaceAll('\n', ' ')
              .replaceAll('\r', ' ');

          bool isArabic = paragraphArabicIdentifier != null &&
              headingContent.contains(paragraphArabicIdentifier);

          TextAlign textAlign = TextAlign.justify;

          if (c.contains('"align":"center"')) {
            textAlign = TextAlign.center;
          } else if (c.contains('"align":"left"')) {
            textAlign = TextAlign.left;
          } else if (c.contains('"align":"right"')) {
            textAlign = TextAlign.right;
          }

          processedParagraphs.add(parseHeadingHTML(headingContent,
              isArabic: isArabic,
              textAlign: textAlign,
              fontFamily: fontFamily,
              arabicFontFamily: arabicFontFamily,
              baseFontSize: fontSize));
        } else if (c.startsWith("paragraph")) {
          String paragraphContent = c
              .substring(
                  c.indexOf("-->") + 3, c.lastIndexOf("<!-- /wp:paragraph -->"))
              .replaceAll('\n', ' ')
              .replaceAll('\r', ' ');

          bool isArabic = paragraphArabicIdentifier != null &&
              paragraphContent.contains(paragraphArabicIdentifier);

          TextAlign textAlign = defaultParagraphTextAlign;

          if (c.contains('"align":"center"')) {
            textAlign = TextAlign.center;
          } else if (c.contains('"align":"left"')) {
            textAlign = TextAlign.left;
          } else if (c.contains('"align":"right"')) {
            textAlign = TextAlign.right;
          }

          processedParagraphs.add(parseParagraphHTML(paragraphContent,
              isArabic: isArabic,
              textAlign: textAlign,
              fontFamily: fontFamily,
              arabicFontFamily: arabicFontFamily,
              baseFontSize: fontSize));
        } else if (c.startsWith("quote")) {
          String paragraphContent = c
              .substring(
                  c.indexOf("-->") + 3, c.lastIndexOf("<!-- /wp:quote -->"))
              .replaceAll('\n', ' ')
              .replaceAll('\r', ' ');

          bool isArabic = paragraphArabicIdentifier != null &&
              paragraphContent.contains(paragraphArabicIdentifier);

          processedParagraphs.add(parseQuoteHTML(paragraphContent,
              isArabic: isArabic,
              textAlign: quoteTextAlignment,
              fontFamily: fontFamily,
              arabicFontFamily: arabicFontFamily,
              baseFontSize: fontSize));
        } else if (c.startsWith("list")) {
          String listContent = c
              .substring(
                  c.indexOf("-->") + 3, c.lastIndexOf("<!-- /wp:list -->"))
              .replaceAll('\n', ' ')
              .replaceAll('\r', ' ')
              .replaceAll("<ol>", "")
              .replaceAll("</ol>", "")
              .replaceAll("</li>", "");

          List<String> listItems = listContent.split("<li>");

          TextAlign textAlign = TextAlign.justify;

          if (c.contains('"align":"center"')) {
            textAlign = TextAlign.center;
          } else if (c.contains('"align":"left"')) {
            textAlign = TextAlign.left;
          } else if (c.contains('"align":"right"')) {
            textAlign = TextAlign.right;
          }

          bool isArabic = paragraphArabicIdentifier != null &&
              listContent.contains(paragraphArabicIdentifier);

          listItems.forEach((li) {
            processedParagraphs.add(parseListHTML(li,
                isArabic: isArabic,
                textAlign: textAlign,
                fontFamily: fontFamily,
                arabicFontFamily: arabicFontFamily,
                baseFontSize: fontSize));
          });
        }
      });
    } catch (exception) {
      print(exception);
    }
    return processedParagraphs;
  }
}

class Thumbnail extends StatelessWidget {
  final String imageURL;
  final Color color;
  final double radius;
  final double aspectRatio;

  const Thumbnail(this.imageURL, this.color, this.radius,
      {this.aspectRatio = 1.7});

  @override
  Widget build(BuildContext context) {
    if (imageURL == null || imageURL.isEmpty || !imageURL.startsWith("http")) {
      return AspectRatio(
        aspectRatio: aspectRatio,
        child: Container(
          color: color,
        ),
      );
    }

    return ClipRRect(
      borderRadius: BorderRadius.all(Radius.circular(radius)),
      child: CachedNetworkImage(
        imageUrl: imageURL,
        fit: BoxFit.cover,
        placeholder: (context, url) {
          return AspectRatio(
            aspectRatio: aspectRatio,
            child: Container(
              color: color,
            ),
          );
        },
      ),
    );
  }
}
